File: variant13_nqueens.py
# Задача о N ферзях для N = 8
# Используем backtracking (обратный ход)

def is_safe(queens, row, col):
    """
    Проверяет, можно ли поставить ферзя в позицию (row, col).

    queens[i] = c  означает, что в строке i стоит ферзь в столбце c.
    """
    for i in range(row):
        # 1) Проверка по столбцу: ферзь уже в этом столбце
        if queens[i] == col:
            return False

        # 2) Проверка по диагоналям:
        # если разница по строке равна разнице по столбцу — одна диагональ
        if abs(queens[i] - col) == abs(i - row):
            return False

    return True


def solve_n_queens(row, n, queens, solutions):
    """
    Рекурсивная функция для расстановки ферзей.

    row       — текущая строка, куда пытаемся поставить ферзя
    n         — размер доски
    queens    — список, где queens[row] = col
    solutions — список всех найденных решений
    """
    # Базовый случай:
    # если row == n, значит ферзи успешно расставлены во всех строках
    if row == n:
        solutions.append(queens[:])  # добавляем копию текущей расстановки
        return

    # Перебираем все возможные столбцы в текущей строке
    for col in range(n):
        if is_safe(queens, row, col):
            # Ставим ферзя
            queens[row] = col

            # Рекурсивный переход к следующей строке
            solve_n_queens(row + 1, n, queens, solutions)

            # Backtracking:
            # убираем ферзя (по сути, отметка, что в этой строке пока ничего нет)
            queens[row] = -1


def print_solution(queens):
    """
    Красивый вывод одной расстановки ферзей в виде доски.
    """
    n = len(queens)
    for row in range(n):
        line = []
        for col in range(n):
            if queens[row] == col:
                line.append("Q")
            else:
                line.append(".")
        print(" ".join(line))
    print()


if __name__ == "__main__":
    n = 8  # размер доски 8x8

    # queens[row] = col — в какой колонке стоит ферзь в строке row
    queens = [-1] * n

    # список для хранения всех решений
    solutions = []

    # Запуск рекурсивного решения
    solve_n_queens(0, n, queens, solutions)

    print(f"Найдено решений: {len(solutions)}\n")

    # Вывод всех решений
    for solution in solutions:
        print_solution(solution)

File: Variant13NQueens.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Variant13NQueens {

    // Проверка, можно ли поставить ферзя в (row, col)
    // queens[i] = c означает: в строке i ферзь в столбце c
    public static boolean isSafe(int[] queens, int row, int col) {
        for (int i = 0; i < row; i++) {
            // 1) Проверка по столбцу
            if (queens[i] == col) {
                return false;
            }
            // 2) Проверка по диагоналям
            if (Math.abs(queens[i] - col) == Math.abs(i - row)) {
                return false;
            }
        }
        return true;
    }

    // Рекурсивный метод для расстановки ферзей
    public static void solveNQueens(int row, int n, int[] queens, List<int[]> solutions) {
        // Базовый случай: все строки обработаны, решение найдено
        if (row == n) {
            solutions.add(queens.clone()); // добавляем копию расстановки
            return;
        }

        // Перебор всех столбцов в текущей строке
        for (int col = 0; col < n; col++) {
            if (isSafe(queens, row, col)) {
                // Ставим ферзя
                queens[row] = col;

                // Рекурсивно переходим к следующей строке
                solveNQueens(row + 1, n, queens, solutions);

                // Backtracking:
                // Опционально можно сбросить позицию
                // queens[row] = -1;
            }
        }
    }

    // Красивый вывод одной доски
    public static void printSolution(int[] queens) {
        int n = queens.length;
        for (int row = 0; row < n; row++) {
            StringBuilder line = new StringBuilder();
            for (int col = 0; col < n; col++) {
                if (queens[row] == col) {
                    line.append("Q ");
                } else {
                    line.append(". ");
                }
            }
            System.out.println(line);
        }
        System.out.println();
    }

    public static void main(String[] args) {
        int n = 8; // размер доски

        // queens[row] = col — ферзь в строке row стоит в столбце col
        int[] queens = new int[n];
        Arrays.fill(queens, -1);

        // Список для хранения всех решений
        List<int[]> solutions = new ArrayList<>();

        // Запускаем backtracking
        solveNQueens(0, n, queens, solutions);

        System.out.println("Найдено решений: " + solutions.size() + "\n");

        // Вывод всех решений
        for (int[] solution : solutions) {
            printSolution(solution);
        }
    }
}
